src/appengine/libs/issue_management/jira/issue_tracker_manager.py

```diff
@@ -0,0 +1,129 @@
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Issue tracker manager functions."""
+from __future__ import absolute_import
+
+from builtins import object
+import jira
+import json
+
+from config import db_config
+
+
+class IssueTrackerManager(object):
+  """Issue tracker manager."""
+
+  def __init__(self, project_name):
+    """"Construct an issue tracker manager instance based on parameters."""
+    self._client = None
+    self.project_name = project_name
+
+  @property
+  def client(self):
+    """HTTP Client."""
+    if self._client is None:
+      self._client = self._create_client()
+
+    return self._client
+
+  def _create_client(self):
+    """Return a client object for querying the issue tracker."""
+    config = db_config.get()
+    credentials = json.loads(config.jira_credentials)
+    jira_url = config.jira_url
+    jira_client = jira.JIRA(
+        jira_url, auth=(credentials['username'], credentials['password']))
+    return jira_client
+
+  def save(self, issue):
+    """Save an issue."""
+    self._update(issue)
+
+  def create(self):
+    """Create an issue."""
+    default_fields = {
+        'project': self.project_name,
+        'summary': 'Default summary',
+        'description': 'Default description',
+        'issuetype': {
+            'name': 'Bug'
+        }
+    }
+
+    jira_issue = self.client.create_issue(fields=default_fields)
+    return jira_issue
+
+  def _update(self, issue):
+    """Update an issue."""
+
+    # Get labels from LabelStore.
+    labels = list(issue.labels)
+
+    # Get components from LabelStore.
+    components = list(issue.components)
+
+    # Get watchers from LabelStore.
+    watchers = list(issue.ccs)
+
+    update_fields = {
+        'summary': issue.title,
+        'description': issue.body,
+        'labels': labels,
+        'components': components,
+    }
+
+    # Only add status if it has changed.
+    # Brittle - we should be pulling the equivalent of 'new' from the policy.
+    if issue.status != 'NOT STARTED':
+      status = {'name': issue.status}
+      update_fields['status'] = status
+
+    if issue.assignee is not None:
+      assignee = {'name': issue.assignee}
+      update_fields['assignee'] = assignee
+
+    # Again brittle - need to pull these strings from policy.
+    if 'Critical - P1' in labels:
```

Comment:
Yes, the best approach here is to define these in the policy config, and save it in the IssueTrackerManager so that we can access it here. 

You can add a field to the jira policy config like so:

```
project_name:
  type: jira
  status:
    assigned: ...
    ...: ...
  priority_labels:
    Critical: 'Critical - P1'
```

etcc

If these are the default, then this PR is OK for now.

Reply:
Those are default priorities from what I can see.

For statuses:
> In the default workflow, issues start as being Open, progressing to In Progress, Resolved and then Closed.

I'll `s/'NOT STARTED'/'Open'` for now... but I want to play with the status update more before I set it to always update. My Jira was giving me errors if I tried to update when there was no change.

This will work for now, but I'll follow up with a PR for a more elegant solution for both of these things.

Reply:
SGTM, thanks!